File: benchmarks/tsb/bench_read_sql.ts
/**
 * Benchmark: readSql — auto-dispatching SQL read (query vs table).
 *
 * `readSql` inspects the first argument: if it looks like a SQL query it
 * calls `readSqlQuery`; otherwise it calls `readSqlTable`. This benchmark
 * exercises both dispatch paths and measures the combined overhead.
 *
 * Dataset: 10 000-row mock adapter (3 columns).
 * Outputs JSON: {"function": "read_sql", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { readSql } from "../../src/index.js";
import type { SqlConnection, SqlResult, SqlRow, SqlValue } from "../../src/index.js";

const ROWS = 10_000;
const WARMUP = 5;
const ITERATIONS = 30;

const columns: string[] = ["id", "value", "label"];
const rows: SqlRow[] = Array.from({ length: ROWS }, (_, i) => ({
  id: i,
  value: Math.round(Math.random() * 1000),
  label: `item_${i % 200}`,
}));

class MockConn implements SqlConnection {
  query(_sql: string, _params?: readonly SqlValue[]): SqlResult {
    return { columns, rows };
  }
  listTables(): readonly string[] {
    return ["sensors"];
  }
}

const conn = new MockConn();

// Two inputs: one that looks like a query, one that looks like a table name.
const SQL_QUERY = "SELECT * FROM sensors";
const TABLE_NAME = "sensors";

function run(): void {
  // Query dispatch path
  const df1 = readSql(SQL_QUERY, conn);
  // Table dispatch path
  const df2 = readSql(TABLE_NAME, conn);
  void df1.shape;
  void df2.shape;
}

for (let i = 0; i < WARMUP; i++) run();

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) run();
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "read_sql",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_read_sql.py
"""
Benchmark: pandas read_sql — auto-dispatching SQL read (query vs table).

Uses sqlite3 as the database backend. The `pd.read_sql` function inspects
the first argument and dispatches to `read_sql_query` or `read_sql_table`
automatically — mirroring tsb's `readSql`.

Dataset: 10 000-row in-memory SQLite table (3 columns).
Outputs JSON: {"function": "read_sql", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import sqlite3
import time

import pandas as pd

ROWS = 10_000
WARMUP = 5
ITERATIONS = 30

# Build an in-memory SQLite database with the same dataset.
con = sqlite3.connect(":memory:")
con.execute(
    "CREATE TABLE sensors (id INTEGER, value INTEGER, label TEXT)"
)
con.executemany(
    "INSERT INTO sensors VALUES (?, ?, ?)",
    [(i, i % 1000, f"item_{i % 200}") for i in range(ROWS)],
)
con.commit()

SQL_QUERY = "SELECT * FROM sensors"
TABLE_NAME = "sensors"


def run():
    # Query dispatch path
    df1 = pd.read_sql(SQL_QUERY, con)
    # Table dispatch path
    df2 = pd.read_sql(TABLE_NAME, con)
    _ = df1.shape
    _ = df2.shape


for _ in range(WARMUP):
    run()

start = time.perf_counter()
for _ in range(ITERATIONS):
    run()
total_ms = (time.perf_counter() - start) * 1000

con.close()

print(
    json.dumps(
        {
            "function": "read_sql",
            "mean_ms": total_ms / ITERATIONS,
            "iterations": ITERATIONS,
            "total_ms": total_ms,
        }
    )
)
